Make Float and Rational compare unequal

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Follows #16924
Fixes #19915 
Fixes #19976 
Fixes #11707
(Probably other issues)

#### Brief description of what is fixed or changed

Previously @smichr's work in #16924 made it so that expressions containing Float would compare unequal to expressions containing Rationals e.g.:
```julia
In [1]: e1 = x**2

In [2]: e2 = x**2.0

In [3]: e1
Out[3]: 
 2
x 

In [4]: e2
Out[4]: 
 2.0
x   

In [5]: e1 == e2
Out[5]: False
```
However #16924 left the behaviour unchanged for individual Number comparisons e.g.:
```julia
In [6]: Float(2.0) == Integer(2)
Out[6]: True
```
This PR extends that change to apply to individual Number so with the PR we have:
```julia
In [1]: Float(2.0) == Integer(2)
Out[1]: False
```
Moreover a Float will only compare equal to another Float if they have the same precision:
```julia
In [3]: Float(2.0, 5) == Float(2.0, 6)
Out[3]: False

In [4]: Float(2.0, 5) == Float(2.0, 5)
Out[4]: True
```
Only a `Float` created with default precision will compare equal to a Python `float`:
```julia
In [5]: Float(2.0, 5) == 2.0
Out[5]: False

In [6]: Float(2.0) == 2.0
Out[6]: True
```

The hash function for Float is changed to return the same as `float.__hash__` if the Float can be represented as a float (e.g. if the exponent is not too large/small):
```julia
In [7]: hash(2.3)
Out[7]: 691752902764107778

In [8]: hash(Float(2.3))
Out[8]: 691752902764107778
```

On master expressions involving Float compare equal but with distinct hashes:
```julia
In [1]: e1 = x**Float(5, 10)

In [2]: e2 = x**Float(5, 20)

In [3]: e1
Out[3]: 
 5.0
x   

In [4]: e2
Out[4]: 
 5.0
x   

In [5]: e1 == e2
Out[5]: True

In [6]: hash(e1)
Out[6]: 1397520323187279713

In [7]: hash(e2)
Out[7]: 7223358913859886252
```
With this PR we get
```julia
In [5]: e1 == e2
Out[5]: False

In [6]: hash(e1)
Out[6]: 5346466018815759790

In [7]: hash(e2)
Out[7]: 5346466018815759790
```
It is not necessary that the hash functions be equal in this example. That's just a consequence of ensuring that they are equal between Float and float. The important thing is that the expressions compare unequal. In that situation there are no constraints on whether the hashes should be equal or not.

This fixes #11707 by ensuring that Floats respect the hash invariant:
```
a == b   -->   hash(a) == hash(b)
```
This also ensures that comparisons using `==` with Floats respect structural equality. Expressions that differ in use of Float/Rational are not structurally equal and the same goes for Floats with differing precision. In this PR those will no longer compare equal.

#### Other comments

This is not yet finished. In particular after making various changes to fix tests I have found that there seems to be a need for some sort of `equalf` function that would compare expressions while ignoring the distinction between Float/Rational or Floats of differing precision.

We could have a function `equalf` so that e.g.:
```
>>> equalf(Float(2.0), 2)
True
>>> equalf(S.Half, 0.5)
True
>>> equalf(x**2.0, x**2)
True
```
There probably also needs to be a function for determining whether or not a Float/Rational represents an integer value. Currently around the codebase we have many cases that do `int(a) == a` which fails in this PR. Instead I have replaced those with `a.round(0) == a` since `a.round(0)` returns a `Float`. With `equalf` you could do `equalf(a, int(a))` but that could be problematic if `a` has a large exponent.

I have also found while working through this that `ordered` fails to break ties between `Float` and `Integer`:
```julia
In [9]: list(ordered([1, 1.0]))
Out[9]: [1, 1.0]

In [10]: list(ordered([1.0, 1]))
Out[10]: [1.0, 1]
```
Probably `sort_key` needs to be fixed for `Float/Integer` but I expect that will lead to a bunch of unrelated test changes so I've left that out of this PR.

Clearly this change is not backwards compatible. I expect that many users will find this more confusing than the current situation but I think it will help with sympy's internal consistency and will reduce bugs in future. For example the bug in #19976 is fixed here:
```julia
In [16]: (2.0*x).replace(lambda e: e.is_Float, Integer)
Out[16]: 2⋅x
```
On master that gives `2.0*x`. I have fixed many other bugs that were ultimately caused by Float comparisons in the past.

One particular point of confusion is that in this PR sympy's Float/Integer no longer behave analogously to Python's float/int:
```julia
In [11]: a, b = 2, 2.0

In [12]: a == b
Out[12]: True

In [13]: S(a) == S(b)
Out[13]: False
```
This also breaks transitivity of `==` although I don't think that's actually a problem any of Python's core data structures in the way that breaking the hash invariant is:
```julia
In [14]: Integer(1) == 1 == 1.0 == Float(1.0) != Integer(1)
Out[14]: True
```

There is one case that I haven't yet changed:
```julia
In [17]: S(0) == S(0.0)
Out[17]: True
```
I'm not sure why that hasn't changed. Possibly it's because `S.Zero` is of a separate class.

I'm putting this up to discuss whether this is wanted and consider whether it is worth a significant break in backwards compatibility.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * BREAKING CHANGE: Float and Integer/Rational no longer compare equal with `a == b`. From now on `Float(2.0) != Integer(2)`. Previously expressions involving `Float` would compare unequal e.g. `x*2.0 != x*2` but an individual `Float` would compare equal to an `Integer`. In SymPy 1.7 a `Float` will always compare unequal to an `Integer` even if they have the same "value". Use `number % 1 == 0` to test if a number has no decimal part (or `not number % 1`); to reject a symbolic expression use `expr.is_Number and not expr % 1`.
    * BREAKING CHANGE: Floats with different prevision no longer compare equal with `a == b`. From now on `Float(2.0, 10) != Float(2.0, 5)`. Previously expressions involving `Float` with different precision would compare equal although they would have different hashes. Now expressions compared with `a == b` involving Floats respect structural equality since Floats of differing precision are not interchangeable.
    * The hash function for Floats and expressions involving Float now respects the hash invariant that if `a == b` then `hash(a) == hash(b)`. This ensures that it is possible to use such expressions in a guaranteed deterministic way in Python's core `set` and `dict` data structures. It is however not recommended to mix up sympy `Basic` objects with non-sympy number types such as core Python's `float` or `int` in the same `set/dict`.
    * It is now possible to replace `Float` with `Integer` in an expression using the `Basic.replace` method.
<!-- END RELEASE NOTES -->

eric-wieser commented:
It worries me a little that this makes the equality non-transitive: `1 == 1.0`, `1.0 == Float(1.0)`, but `Float(1.0) != 1`.

oscarbenjamin commented:
> this makes the equality non-transitive

The problem is that core Python has this as a broken feature:
```julia
In [5]: {1, 1.0}                                                                                                                                              
Out[5]: {1}

In [6]: {1.0, 1}                                                                                                                                              
Out[6]: {1.0}
```
There's nothing we can do about that in sympy and it's too late to fix in core Python. Because `==` and `__eq__` are directly tied to the fundamental set/dict data structures we need them to be able to distinguish between `Float` and `Rational`. Core Python could have chosen some other way such as not having float and int compare equal or not using the same equality operator for sets and dicts as for `==` etc. It's all too late now but we can't afford to make the same mistake in sympy where distinguishing Float and Rational is essential.

The transitivity issue will only arise when mixing sympy and non-sympy types involving floats in the same set/dict. As you can see above though mixing float and int with non-sympy types alone is already broken. With this PR we can at least be internally consistent with sympy types.

asmeurer commented:
This would definitely fix the issues in question. My biggest concern here is how this would break user code. Seeing how it breaks SymPy should give us an idea. 

If this equalf function ends up being used everywhere, that may be a sign that this is really a bad idea. Currently there are a lot of cases in SymPy where an operation will work just fine for an expression with rational coefficients but not if they are replaced with floating point coefficients, and I worry that this change will only make that situation worse.

eric-wieser commented on sympy/core/tests/test_evalf.py:
```diff
@@ -317,11 +317,11 @@ def test_evalf_power_subs_bugs():
-    assert exp(x).evalf(subs={x: 0}) == 1
-    assert ((2 + I)**x).evalf(subs={x: 0}) == 1
-    assert (0**x).evalf(subs={x: 0}) == 1
+    assert (x**x).evalf(subs={x: 0}) == 1.0
+    assert (3**x).evalf(subs={x: 0}) == 1.0
+    assert exp(x).evalf(subs={x: 0}) == 1.0
+    assert ((2 + I)**x).evalf(subs={x: 0}) == 1.0
+    assert (0**x).evalf(subs={x: 0}) == 1.0
```
Would it be possible (or sensible) to lessen the pain so that `Float(1.0) == 1` even thought `Float(1.0) != Rational(1)`?

oscarbenjamin replied:
Does that lessen the pain? I think it would add to the confusion.

I've tried to preserve the behaviour that if `x` is a sympy object then `x == y` iff `x == sympify(y)`. Implicit sympification should make it so that Float/float are interchangeable and also int/Integer are interchangeable provided that some sympy type is involved somewhere.

asmeurer commented:
> The hash function for Float is changed to return the same as float.__hash__ if the Float can be represented as a float (e.g. if the exponent is not too large/small):

Maybe I'm missing it, but it doesn't look like your code handles the case where the exponent is small. `float(Float("1e-400"))` returns `0.0`.

oscarbenjamin commented:
> it doesn't look like your code handles the case where the exponent is small.

You're right it doesn't. I'll fix that...

oscarbenjamin commented:
> My biggest concern here is how this would break user code. Seeing how it breaks SymPy should give us an idea.

It broke lots of tests and quite a few bits of main code needed changing (in fairly simple ways). Some of the changes were not easy to debug.

> Currently there are a lot of cases in SymPy where an operation will work just fine for an expression with rational coefficients but not if they are replaced with floating point coefficients

Many parts of sympy have not been written to handle Floats but can sometimes work with them partly by accident because of a haphazard mix of implicit and explicit conversions/compensations that are dotted around.

Probably this PR would introduce new problems with Floats in some of those cases. However it could also make it cleaner to fix them properly because handling of Floats would need to be more explicit.

coproc commented:
> The problem is that core Python has this as a broken feature:
> 
> ```julia
> In [5]: {1, 1.0}                                                                                                                                              
> Out[5]: {1}
> 
> In [6]: {1.0, 1}                                                                                                                                              
> Out[6]: {1.0}
> ```

Is this really a broken feature? In Python we have `1.0 == 1` and neither is a canonical representation of the other.

coproc commented:
> It worries me a little that this makes the equality non-transitive: `1 == 1.0`, `1.0 == Float(1.0)`, but `Float(1.0) != 1`.

To keep transitivity we would have to give up `1.0 == Float(1.0)`. I think this is reasonable since sympy is about being exact and, for instance, `2.0 ** 2000` overflows whereas `Float(2.0)**2000` does not. When mixing floats and sympy types, like in `1.0*x`, one could consider/declare a float like `1.0` as a shorthand notation for `Float(1.0)` (which is what happens internally anyway) - and the rest of this pull request would still be/make things consistent, i.e. `Float(1) != Integer(1)` and `Float(1) != Float(1,1)`. Also `Float(0) != Integer(0)` should hold then.

oscarbenjamin commented:
> Is this really a broken feature? In Python we have `1.0 == 1` and neither is a canonical representation of the other.

The fact that `1.0` and `1` are indistinguishable in sets and dicts is a broken feature because these are not interchangeable.

coproc commented:
> > Is this really a broken feature? In Python we have `1.0 == 1` and neither is a canonical representation of the other.
> 
> The fact that `1.0` and `1` are indistinguishable in sets and dicts is a broken feature because these are not interchangeable.

Sure, `1.0` and `1` are not interchangeable in all situations (e.g. `1.0 + 2**53 != 1 + 2**53`), but then already `1.0 == 1` must be considered as broken.

coproc commented:
> The fact that `1.0` and `1` are indistinguishable in sets and dicts is a broken feature because these are not interchangeable.

@oscarbenjamin I am still trying to understand why you consider the behaviour of `1.0` and `1` with respect to sets and dicts special ("broken"). I do not see any difference to the behaviour with respect to lists:
```python
>>> 1 in {1.0}
True
>>> 1 in [1.0]
True
>>> {1} == {1.0}
True
>>> [1] == [1.0]
True
```
Everything seems to boil down to `1 == 1.0`, which may be useful for everyday programming, but not for exact computations.

So I also want to question whether making `1.0 == Float(1.0)` is really so helpful that the broken transitivity (as mentioned above) should be ignored. A standard practice when programming with floats is to avoid checking for equality anyway (and to look at the absolute difference instead).

I am very much in favour of all the other changes/achievements of this pull request.

oscarbenjamin commented:
The difference with lists is that the list does not interchange `1` and `1.0`:
```julia
In [1]: s = {1.0}

In [2]: s.add(1)

In [3]: s
Out[3]: {1.0}

In [4]: l = [1.0]

In [5]: l.append(1)

In [6]: l
Out[6]: [1.0, 1]
```
From a language design perspective a decision was made to have `1 == 1.0` and also a decision was made that the behaviour of dicts and sets would be based on whether `A == B`. Note that dicts and sets could have been based on some other method (they already have their own special `__hash__` method) so the fact that `1 == 1.0` does not have to mean that `1` and `1.0` are interchangeable in sets and dicts.

From sympy's perspective I approach thinking about this in reverse: `1` and `1.0` should not be interchangeable but Python's fundamental data structures will make them interchangeable if `1 == 1.0` so therefore we need to have `1 != 1.0`.

> So I also want to question whether making `1.0 == Float(1.0)` is really so helpful that the broken transitivity (as mentioned above) should be ignored.

We generally want to have `a == sympify(a)` so that users can do things like:
```julia
In [7]: x = Symbol('x')

In [8]: f = 1.0

In [9]: x + f - x == f
Out[9]: True
```
The transitivity issue isn't ignored. It just only occurs when mixing sympy and non-sympy types in the same data structure which is in any case a bad practice to be avoided.

coproc commented:
> We generally want to have `a == sympify(a)` so that [...]

Ok, so let me bring things together. These are the conditions currently involved:

1. seemless embedding of Python types into sympy: `a == sympify(a)`
1. transitivity: `a == b` and `b == c` imply `a == c`, or as one Python expression: `not (a == b and b == c) or a == c`
1. distinction of exact and approximate results in sympy: `Integer(n) != Float(n)`

Now the problem is that all three conditions cannot be fulfilled at the same time: 

- from Python we have `1 == 1.0`
- from (1) seemless embedding we have `1 == Integer(1)` and `1.0 == Float(1)`
- from (2) transitivity (and to be precise: symmetry of `==`) we get `Integer(1) == Float(1)`
- but this contradicts (3) distinction of exact and approximate results in sympy: `Integer(1) != Float(1)`

This pull request fulfills (1) and (3), but breaks (2) when also considering the builtin non-sympy types `int` and `float`.

oscarbenjamin commented:
> This pull request fulfills (1) and (3), but breaks (2) when also considering the builtin non-sympy types `int` and `float`.

Yes, exactly. However transitivity is preserved among sympy types. With this PR everything involving `==`, `set`, and `dict` and sympy types is internally consistent and matches the expected rules.

asmeurer commented:
This is not directly part of this change, but related to the transitivity change. I noticed this behavior:

```py
>>> Integer(0) == False
True
>>> S.false == 0
False
>>> S.false == Integer(0)
False
```

The first one should also give False (even though `0 == False` gives True). 

Of course, integers and booleans not being equal more stomachable than integers and floats not being equal, but I also think that non-transitivity isn't a huge deal in practice.

oscarbenjamin commented:
You have to remember that `a == b` is not for mathematical equality: it is for structural equality. Structural equality should mean that `a` and `b` are entirely equivalent expressions and that we would be happy to interchange them. In Python `a == b` makes the objects `a` and `b` interchangeable in sets and dicts etc. We don't want Float and Integer to be interchangeable which is why the behaviour in #19976 is considered a bug (caused by code that assumes `a == b` means `a` and `b` are interchangeable).

Mathematical equality is tested with `Eq(a, b)` or similar e.g.:
```julia
In [103]: 2*x == 2.0*x                                                                                                                         
Out[103]: False

In [104]: Eq(2*x, 2.0*x)                                                                                                                       
Out[104]: True
```
Also as these examples show Float and Integer already compare unequal with `==` when they form any part of a larger expression since #16924. It is only the case of a plain Float that compares equal to a plain Integer.

asmeurer commented:
I've seen a lot of SymPy code in the wild, from places like Twitter and StackOverflow, and I can say that a very large fraction of users use floating point constants in their expressions. In most cases, I don't think that users are using floating point numbers because the user believes they are doing "approximate calculations". Rather, they are used because

- They are seen as just a "decimal number". To most people, decimal numbers are no different than rational or integral numbers (and mathematically, that is correct, but obviously even for finite decimals this has problems because floats use base-2).
- They represent some real-world data that is being combined with an expression somehow.
- They don't realize that things like `1/2` create a floating point number, and the implications of that. 

In general, I'd say a small percentage of SymPy users actually fully grasp the concept of a floating point number, and the implications of using them. Even people who work with numeric code and are familiar with floating point numbers aren't likely to understand the implications of how SymPy handles them. The idea that SymPy actually works better with rational constants than with floating point constants is very counterintuitive to most people. 

I think we can can do about this is

- Document this better. Particularly the places where they can cause real issues, e.g., floats as coefficients are generally not so bad, but floats a exponents can be really bad. 
- Try to make SymPy work better with floating point numbers. This means that certain core algorithms should be made to work with floats (particularly in the polys). And in places where they don't work with floats, we can experiment with converting to rational first (although `solve` has had this with the `rational=True` flag, and it has caused a lot of problems over the years). 

There are way too many cases where some function will not work at all with an expression with floating point constants, but will work just fine if the floating point constants are replaced with rationals. In some cases, this is understandable. For example, `x**2.1` is read as `x**(4728779608739021/2251799813685248)`, rather than the expected `x**(21/10)`, which can mean working with a degree 2251799813685248 polynomial rather than a degree 10 polynomial. But in other cases, there's really no good reason for it, and even `nsimplify(expr, rational_conversion='exact')` works. It just comes down to some core algorithms not handling floats well at all.

What does this all mean for this pull request? I'm not sure. I would expect if anything this would make things worse, because places that would currently handle floats via something like `if n == 1` would stop handling them, and make even more cases where floats unnecessarily don't work. So my gut feeling is that a change like this is not a good idea.

oscarbenjamin commented:
I think most users would be happier if it was the other way round and all floats sympified to Rational (using nsimplify) and only explicitly created Floats were actually treated as Float by sympy.

coproc commented:
> You have to remember that `a == b` is not for mathematical equality: it is for structural equality. [..]
> Mathematical equality is tested with `Eq(a, b)` or similar e.g.:

Thanks for pointing out. I remember now having seen `Eq` before, but it is tempting to fall back to `==` ...

jksuom commented:
> So my gut feeling is that a change like this is not a good idea.

That was also my first reaction but I'm not so sure any more as the tests have passed. A new method would have to be created for testing numerical equality. The current `__eq__` methods in `numbers` could be changed to `eq`, so `n == 1` would become `n.eq(1)` (and similarly for `__ne__`).

> all floats sympified to Rational

That looks feasible to me. The conversion could be done so that the denominator is (a factor of) a power of 10 instead of a high power of 2 (cf. https://github.com/sympy/sympy/issues/17648#issuecomment-536286768).

czgdp1807 commented:
@oscarbenjamin @asmeurer @jksuom Any further discussions needed on this?

oscarbenjamin commented:
I think it would be useful to take some of the changes from this PR e.g. the changes to tests and apply them anyway even if the main change to comparisons is not merged. I also think it would be good to make helper functions for comparing expressions with mixed Float/Rational and use them in the places highlighted in this PR. We can do these things to extract most of the code changes from this PR without making any backward incompatible change to comparisons right now.

The main change `Float != Rational` is not backwards compatible and I think would possibly need to be part of a release like a sympy 2.0 that was explicitly intended to be incompatible. That's something that should be planned carefully over a long time though so it's not just a case of merging this now.

The change to float comparison here improves sympy's internal consistency at the expense of making things a bit more awkward for users. I think that we can dramatically improve the user experience (for most users) by having float sympify to Rational but that's a really big incompatible change. If we did that then it would alleviate the problems caused by the change to Float/Rational comparison in this PR so it would make sense for those changes to go together. I think it would be good to do that eventually but it would definitely need to be part of a major breaking release.

It might be useful to have the test suite run also with these settings enabled to pick up on any sloppy tests and checks that mix Float/Rational in the codebase. Also if we do want to make significant incompatible changes like this then it is good to provide a way of enabling the changes in advance (the equivalent of a `__future__` import) so that downstream libraries can prepare for the changes and can more easily bridge multiple versions.

asmeurer commented:
It would be a lot of work, so I don't blame anyone if no one wants to actually do it, but it would be interesting to compare this side-by-side with second pull request implementing the alternative idea of making floats and rationals always compare equal. This would probably be easier than this PR since Float == Rational already returns True when they are exactly equal, but it would also mean removing workarounds like `_aresame` and `x**1.0` to make this consistent throughout the codebase, and also fixing the hashing issue #11707.

oscarbenjamin commented:
I moved the trivial test changes from here to #23499

oscarbenjamin commented:
I've scanned through the diff here to do a survey of places in the code that need to compare floats and these are the counts of things used:

Many places want to mix integers, rationals and floats and consider integer valued floats equivalent to integers and in some cases half integer floats as being equal to rational numbers. Out of these cases there are 4 where a float is expected but 42 where the expectation is that the input might be a float or might be a rational and we want to consider those equivalent if they represent the same number.

- 12 calls: check if `1` or `1.0`
- 10 calls: check if `-1` or `-1.0`
- 8 calls: check if integer-valued
- 7 calls: check if `1/2` or `0.5`
- 4 calls: check if `0` or `0.0`
- 3 calls: check if `-1/2` or `-0.5`
- 1 call: compare `x = y` considering floats and integers equal
- 1 call: check if `2` or `2.0`

Most of these could be served by a function called say `equal_valued` that compares to numbers and considers them equivalent if they have the exact same numeric value either because:
- they are equal rationals
- they are floats with equal value
- mixed float/rational but exact value is equal

In master these are all satisfied just with `x == y` but here they are not for two reasons:
- Rational and Float always compare unequal
- Floats with mixed precision compare unequal

It is not a good idea in general to simply convert both operands to float or both to rational. Converting a float to a rational can be a very expensive operation for floats with large exponent. Converting rational to float is inexact and in any case floats with different precision will still compare unequal so we would need to change the precision of the floats in some cases. It is better though to compare them without conversion.

A function `equal_valued` could be:
```python
def equal_valued(x, y):
    """Compare expressions treating a rational as equal to a float"""
    x = _sympify(x)
    y = _sympify(y)

    # Handle everything except Float/Rational first
    if not x.is_Float and not y.is_Float:
        return x == y
    elif x.is_Float and y.is_Float:
        # Compare values without regard for precision
        return x._mpf_ == y._mpf_
    elif x.is_Float:
        x, y = y, x
    if not x.is_Rational:
        return False

    # Now y is Float and x is Rational
    sign, man, exp, _ = y._mpf_
    if sign:
        man = -man
    if man != x.p:
        return False

    # x = p/q  and  y = man*2**exp  and  p == man.
    # Now if exp >= 0 y is an integer but integer x means q == 1.
    if exp > 0:
        return False
    elif exp == 0:
        return x.q == 1
    else:
        exp = -exp
        return x.q.bit_length() - 1 == exp and (1 << exp) == x.q
```
With that the various cases above are just `equal_valued(x, 1)`, `equal_valued(x, 0.5)` or `equal_valued(x, y)` etc. This `equalf` function does not recursively walk the expression trees looking for floats and rationals that might compare equal because it is really just intended for the case where both expressions are expected to be numbers and most of the time one argument is guaranteed to be an integer anyway.

I think that there is a need to have a function that does recurse the expression trees e.g. for the test suite to compare expressions with floats. That should have other options though such as a tolerance for comparing floats that are not exactly equal. The intention of `equal_valued` above is that it is still an exact value comparison but just one that admits floats as representing their exact value.

What isn't handled by `equal_valued` are the 8 places that want to check for an Integer or an integer-valued Float. This is a bit trickier though because all large floats are integers so we probably need something that takes into account precision. Not taking into account precision is what gives e.g. this:
```python
In [48]: sin(1.132e100*pi)
Out[48]: 0
```

oscarbenjamin committed:
make float compare unequal with rational

oscarbenjamin commented:
I've opened https://github.com/sympy/sympy/pull/24625 to extract the `equal_valued` function from here.

oscarbenjamin committed:
change doctest...

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -2071,18 +2071,21 @@ def _aresame(a, b):
 
-    This routine was written to provide a query for such cases that
-    would give false when the types do not match:
+    Before SymPy 1.7 the two comparisons above would have given True rather
```
I'd forgotten how long this PR has been open!

smichr committed:
Merge branch 'master' into pr_float_equality

smichr committed:
Update numbers.py

smichr commented:
Python's errors need not be maintained in SymPy. This is the right direction regardless of the pain. Unless there are objections I will commit this.

smichr committed:
Update test_integrals.py

oscarbenjamin commented:
Some of the changes needed are like using `x == x.round(0)` show that really a helper function is needed. We should add something like `.equalf` as a method for checking if two expressions are equal modulo float vs rational and use that here before merging it.

If the changes here look awkward then we have to assume that it would also be awkward for downstream code so we should try to mitigate that before making any change in behaviour. I suggest adding e.g. `.equalf` in a separate PR and then once it is merged updating this PR to use it. Note that `.equalf` is already useful for larger expressions containing floats since those already compare unequal with `==`.

smichr commented:
`equalf` is pretty close to `equals` -- what about `eq`?
```python
def _eq(x, y):
    if not isinstance(x.func, type(y)):
        return False
    if len(x.args) != len(y.args):
        return False
    if x.args:
        return all(i._eq(j) for i, j in zip(x.args, y.args))
    return equal_valued(x, y)

def eq(self, other):
    """return True if self and other are the same except for
    precision in the numbers contained in each.

    >>> from sympy import Float, Rational
    >>> from sympy.abc import x
    >>> Float(2) == 2.0
    True
    >>> Float(2) == 2
    False
    >>> x**2 == x**2.0
    False
    >>> Float(2) == Rational(2)
    False
    >>> Float(2).eq(Rational(2))
    True
    >>> Float(2, 10).eq(Float(2, 20))
    True
    >>> (x**2).eq(x**2.0)
    True
    """
    other = sympify(other)
    return (
        isinstance(other.func, type(self)) and
        len(self.args) == len(other.args) and
        all(
            _eq(i,j) for i,j in zip(self.args or [self],
                                   other.args or [other])))
```

oscarbenjamin commented:
> `equalf` is pretty close to `equals` -- what about `eq`?

Perhaps `equalf` is too close to `equals` but I think that `eq` is too short and does not convey its meaning. In context if we already have `==` and `.equals` which people already find confusing adding `eq` to the mix with a name that gives no sense of how it differs from the others might add to the confusion.

The intention of `equalf` is that it would still be for structural equality like `==` apart from handling of floats with differing precision and distinguishing float from rational. That's very different from `equals` which tests if two expressions are mathematically equivalent regardless of structure. Probably the name should be very clearly distinct from `equals` but also the name should convey its intent when used.

Also we should probably want to have something like `equalf_approx` for comparing if two expressions are close given small floating point errors. Then if we had that we would want `equalf` to be called something like `equalf_exact` or perhaps `equalf_value`.

smichr commented:
I see there is some precedent for `feq` and `fne` (in the FORTRAN context).

I would suggest a tweak to the number model, however, by only blessing **default Float**/Rational equality so `eq(Float(.5), S.Half)` but `not eq(Float(.5,10), S.Half)` and also allow Float(.6) to compare equal to 6/10. While the undrlying binary representation is an easy test, it is of less concern -- the reason we are allowing .5 to compare with 1/2 is that that is the way someone would type the number. The fact that it happens to be exact, except for precision in binary, is not of interest to the average user who would be surprised to have .6 not be equal to 6/10.

asmeurer commented:
I would suggest for `approx` to be a keyword argument to the method. Although it's unclear to me if users would ever actually need that or not. I would be OK with omitting it for now unless it's also needed for some library code.

As far as the name, I'm actually more worried that `equalf` looks too much like `evalf` (things are already hard enough with simplify and sympify).

smichr commented:
Here are some types of comparisons:

1. exact  (type and precision matter)
2. binary equal ignore prec for int -> float(1) eq 1, Float(1, 10) == Float(1, 15)
3. binary equal ignore precision -> float(.5) eq Rational(1, 2)
4.  nearest fp prec with fiddling -> Rational(str(.6)) eq Rational(3, 5)
6. nearest 10**-3 -> Rational(.6).limit_denominator(10**3) eq Rational(3, 5)

Looking at your diff, I think we can just stay out of the weeds with testing floats as ints (that is how the round(0) is being used) by testing for `int(a) == float(a)` like `not a%1` (and sympy_number%1 -> Integer, and float%1 -> float which, when 0.0, is equal to 0). To test if `(a + b + c)` is an integer test `(a + b + c)%1==0`. 

It is faster than `round`:
```python
>>> r = [Rational(random.random()*1000).round(0) for i in range(1000)]  # Float
>>> t=time();ok=[i%1 for i in r];time()-t
0.031199932098388672
>>> t=time();ok=[i.round(0) for i in r];time()-t
0.6708011627197266

>>> r = [Rational(random.random()) for i in range(1000)]  # Rational
>>> t=time();ok=[i%1 for i in r];time()-t
0.23400020599365234
>>> t=time();ok=[i.round(0) for i in r];time()-t
2.6364047527313232

>>> r = [Integer(i) for i in range(100000)]  # 100x more Integers
>>> t=time();ok=[i%1 for i in r];time()-t
0.2652006149291992
>>> t=time();ok=[i.round(0) for i in r];time()-t
2.511604070663452
```
Within expressions, it seems that users mostly need to convert floats to Rational, not compare it to a similar expression with Rational replacements.

oscarbenjamin commented:
> Looking at your diff, I think we can just stay out of the weeds with testing floats as ints (that is how the round(0) is being used) by testing for `int(a) == float(a)`

This test might not be correct for large values because both `float` and `int` can do some rounding.

I agree though that actually the number one thing needed for the diff here is a function to test if an expression is either an `Integer` or an integer-valued `Float`. We can't use `is_integer` for this because it always returns False for Floats (perhaps it should be `None`) but clearly there are many parts of the codebase that want to have a test like this.

smichr commented:
> by testing for `int(a) == float(a)`

This was an imperfect way of me trying to say: test whether a number has no decimal portion (i.e. would pass `as_int(a)`).

smichr commented on sympy/holonomic/holonomic.py:
```diff
@@ -1462,7 +1462,7 @@ def _frobenius(self, lb=True):
                 continue
             for j in grp:
-                if int(j[0] - i) == j[0] - i:
+                if (j[0] - i).round(0) == j[0] - i:
```
```suggestion
                if not (j[0] - i) % 1:
```

smichr commented on sympy/core/numbers.py:
```diff
@@ -1038,7 +1038,7 @@ def __mod__(self, other):
         if isinstance(other, Float) and global_parameters.evaluate:
             r = self/other
-            if r == int(r):
+            if r == r.round(0):
```
```suggestion
            if not r % 1:
```

smichr commented on sympy/holonomic/holonomic.py:
```diff
@@ -1474,7 +1474,7 @@ def _frobenius(self, lb=True):
 
         allpos = all(i >= 0 for i in reals)
-        allint = all(int(i) == i for i in reals)
+        allint = all(int(i) == i or i.round(0) == i for i in reals)
```
```suggestion
        allint = not any(i % 1 for i in reals)
```

smichr commented on sympy/holonomic/holonomic.py:
```diff
@@ -1964,7 +1964,7 @@ def to_hyper(self, as_list=False, _recur=None):
             for j, i in enumerate(nonzeroterms):
 
-                if i < 0 or int(i) != i:
+                if i < 0 or not (int(i) == i or i.round(0) == i):
```
```suggestion
                if i < 0 or i % 1:
```

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -602,10 +602,10 @@ def isprime(n):
     False
     >>> n = Float(near_int, 10)  # truncated by precision
-    >>> n == int(n)
+    >>> n == n.round(0)
```
```suggestion
    >>> n % 1 == 0
```

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -602,10 +602,10 @@ def isprime(n):
     True
     >>> n = Float(near_int, 20)
-    >>> n == int(n)
+    >>> n == n.round(0)
```
```suggestion
    >>> n % 1 == 0
```

smichr commented on sympy/physics/wigner.py:
```diff
@@ -336,11 +336,11 @@ def _big_delta_coeff(aa, bb, cc, prec=None):
     """
 
-    if int(aa + bb - cc) != (aa + bb - cc):
+    if int(aa + bb - cc) != (aa + bb - cc) and (aa + bb - cc).round(0) != (aa + bb - cc):
```
```suggestion
    if (aa + bb - cc) % 1:
```

smichr commented on sympy/physics/wigner.py:
```diff
@@ -336,11 +336,11 @@ def _big_delta_coeff(aa, bb, cc, prec=None):
+    if int(aa + bb - cc) != (aa + bb - cc) and (aa + bb - cc).round(0) != (aa + bb - cc):
         raise ValueError("j values must be integer or half integer and fulfill the triangle relation")
-    if int(aa + cc - bb) != (aa + cc - bb):
+    if int(aa + cc - bb) != (aa + cc - bb) and (aa + cc - bb).round(0) != (aa + cc - bb):
```
```suggestion
    if (aa + cc - bb) % 1:
```

smichr commented on sympy/physics/wigner.py:
```diff
@@ -336,11 +336,11 @@ def _big_delta_coeff(aa, bb, cc, prec=None):
+    if int(aa + cc - bb) != (aa + cc - bb) and (aa + cc - bb).round(0) != (aa + cc - bb):
         raise ValueError("j values must be integer or half integer and fulfill the triangle relation")
-    if int(bb + cc - aa) != (bb + cc - aa):
+    if int(bb + cc - aa) != (bb + cc - aa) and (bb + cc - aa).round(0) != (bb + cc - aa):
```
```suggestion
    if (bb + cc - aa) % 1:
```

smichr commented on sympy/polys/domains/gmpyintegerring.py:
```diff
@@ -35,7 +35,7 @@ def from_sympy(self, a):
         if a.is_Integer:
             return GMPYInteger(a.p)
-        elif a.is_Float and int(a) == a:
+        elif a.is_Float and a.round(0) == a:
```
```suggestion
        elif a.is_Float and not a % 1:
```

smichr commented on sympy/polys/domains/integerring.py:
```diff
@@ -69,7 +69,7 @@ def from_sympy(self, a):
         if a.is_Integer:
             return MPZ(a.p)
-        elif a.is_Float and int(a) == a:
+        elif a.is_Float and a.round(0) == a:
```
```suggestion
        elif a.is_Float and not a % 1:
```

smichr commented on sympy/polys/domains/pythonintegerring.py:
```diff
@@ -33,7 +33,7 @@ def from_sympy(self, a):
         if a.is_Integer:
             return PythonInteger(a.p)
-        elif a.is_Float and int(a) == a:
+        elif a.is_Float and a == a.round(0):
```
```suggestion
        elif a.is_Float and not a % 1:
```

smichr commented on sympy/utilities/misc.py:
```diff
@@ -560,6 +560,6 @@ def as_int(n, strict=True):
             result = int(n)
         except TypeError:
             raise ValueError('%s is not an integer' % (n,))
```
```suggestion
            if n % 1:
                raise TypeError
        except TypeError:
            raise ValueError('%s is not an integer' % (n,))
```

smichr commented on sympy/utilities/misc.py:
```diff
@@ -560,6 +560,6 @@ def as_int(n, strict=True):
         except TypeError:
             raise ValueError('%s is not an integer' % (n,))
-        if n != result:
+        if not (n == result or n == n.round(0)):
             raise ValueError('%s is not an integer' % (n,))
         return result
```
```suggestion
        return int(n)
```

smichr committed:
use %1 to test for decimal portion in number

smichr committed:
remove _aresame and _ne (now unused)

smichr commented on sympy/utilities/misc.py:
```diff
@@ -557,9 +557,8 @@ def as_int(n, strict=True):
     else:
         try:
-            result = int(n)
+            if n % 1:
+                raise TypeError
```
```suggestion
            result = int(n)
```

smichr committed:
Update misc.py

smichr committed:
edit _aresame docstring

smichr commented on sympy/core/basic.py:
```diff
@@ -2113,20 +2113,21 @@ def _aresame(a, b):
+    class types.
 
-    >>> from sympy.core.basic import _aresame
+    >>> eye(1) == S(eye(1))  # mutable vs immutable
```
```suggestion
    >>> from sympy.core.basic import _aresame
    >>> eye(1) == S(eye(1))  # mutable vs immutable
```

smichr committed:
Update sympy/core/basic.py

smichr commented on sympy/core/tests/test_numbers.py:
```diff
@@ -204,10 +205,11 @@ def test_divmod():
-    assert divmod(-S(3.5), S(-2)) == divmod(-3.5, -2)
-    assert divmod(S(0.0), S(9)) == divmod(0.0, 9)
-    assert divmod(S(0), S(9.0)) == divmod(0, 9.0)
+    # sympy's divmod gives an Integer for the quotient rather than a float
+    #assert divmod(S(3.5), S(-2)) == divmod(3.5, -2)
+    #assert divmod(-S(3.5), S(-2)) == divmod(-3.5, -2)
+    #assert divmod(S(0.0), S(9)) == divmod(0.0, 9)
+    #assert divmod(S(0), S(9.0)) == divmod(0, 9.0)
```
```suggestion
    # sympy's divmod gives an Integer for the quotient rather than a float
    dmod = lambda a, b: tuple([j if i else int(j) for i, j in enumerate(divmod(a, b))])
    assert divmod(S(3.5), S(-2)) == dmod(3.5, -2)
    assert divmod(-S(3.5), S(-2)) == dmod(-3.5, -2)
    assert divmod(S(0.0), S(9)) == dmod(0.0, 9)
    assert divmod(S(0), S(9.0)) == dmod(0, 9.0)
```

smichr commented on sympy/core/tests/test_numbers.py:
```diff
@@ -190,7 +190,8 @@ def test_divmod():
     assert divmod(S(-3), 2) == (-2, 1)
 
     assert divmod(S(4), S(-3.1)) == Tuple(-2, -2.2)
```
```suggestion
```

smichr commented on sympy/core/tests/test_numbers.py:
```diff
@@ -190,7 +190,8 @@ def test_divmod():
 
     assert divmod(S(4), S(-3.1)) == Tuple(-2, -2.2)
-    assert divmod(S(4), S(-2.1)) == divmod(4, -2.1)
+    # divmod(4, -2.1) gives an inaccurate result for the remainder
+    # assert divmod(S(4), S(-2.1)) == divmod(4, -2.1)
     assert divmod(S(-8), S(-2.5) ) == Tuple(3, -0.5)
```
```suggestion
```

smichr committed:
update divmod tests

smichr commented on sympy/core/basic.py:
```diff
@@ -2113,20 +2113,22 @@ def _aresame(a, b):
-    This routine was written to provide a query for such cases that
-    would give false when the types do not match:
+    However, other object might compare the same. This
+    routine will only return True if two expressions identical in terms of
+    class types.
```
```suggestion
    routine will only return True if two expressions are identical in
    terms of class types.
```

smichr committed:
edit docstring

smichr committed:
align divmod,floordiv with python for other = -int

smichr commented:
The divmod "alternate" answer for `divmod(4,-2.0)` exists in master, too, giving (-3, -2.0) instead of (-3, 0); it should actually be `(-3, -0.0)` but SymPy doesn't have signed zero, so just `(-3, 0.0)`.

smichr commented on sympy/core/numbers.py:
```diff
@@ -370,7 +370,9 @@ def __divmod__(self, other):
+            w = int(rat)
+            if rat < 0 and other % 1:  # correct if negative and not int
+                w -= 1
             r = self - other*w
```
```suggestion
            r = self - other*w
            if not r and (isinstance(self, Float) or isinstance(other, Float)):
                r = Float(r)  # python has signed 0, but not SymPy

```

smichr committed:
return 0.0, not 0, if Float in divmod args

smichr committed:
Update test_numbers.py

smichr commented on sympy/core/numbers.py:
```diff
@@ -370,8 +370,13 @@ def __divmod__(self, other):
+            if rat < 0 and other % 1:  # correct if negative and not int
+                w -= 1
             r = self - other*w
+            if not r and (isinstance(self, Float) or isinstance(other, Float)):
```
```suggestion
            if isinstance(self, Float) or isinstance(other, Float):
```

smichr commented on sympy/core/numbers.py:
```diff
@@ -370,8 +370,13 @@ def __divmod__(self, other):
+                w -= 1
             r = self - other*w
+            if not r and (isinstance(self, Float) or isinstance(other, Float)):
+                r = Float(r)  # python has signed 0, but not SymPy
```
```suggestion
                r = Float(r)  # in case w or r is 0
```

smichr committed:
r is Float if either input is Float

smichr committed:
Update test_numbers.py

smichr committed:
fix alignment of divmod

smichr commented on sympy/core/tests/test_numbers.py:
```diff
@@ -204,10 +201,11 @@ def test_divmod():
+    dmod = lambda a, b: tuple([j if i else int(j) for i, j in enumerate(divmod(a, b))])
+    for a in (4, 4., 4.25, 0, 0., -4, -4. -4.25):
+        for b in (2, 2., 2.5, -2, -2., -2.5):
+            assert divmod(S(a), S(b)) == dmod(a, b)
```
```suggestion
            assert divmod(S(a), S(b)) == dmod(a, b)
    # done with Float/Integer now Rational
    for a in range(-20, 21, 10):
           a = a/S(5)
           for b in (-10, -5, 5, 10):
                  b = b/S(5)
                  assert __divmod__(a, b) == dmod(a, b)
```

smichr committed:
Update sympy/core/tests/test_numbers.py

smichr commented on sympy/core/tests/test_numbers.py:
```diff
@@ -204,10 +201,17 @@ def test_divmod():
+           a = a/S(5)
+           for b in (-10, -5, 5, 10):
+                  b = b/S(5)
+                  assert __divmod__(a, b) == dmod(a, b)
```
```suggestion
                  assert divmod(a, b) == dmod(a, b)
```

smichr commented on sympy/core/tests/test_numbers.py:
```diff
@@ -204,10 +201,17 @@ def test_divmod():
+    for a in (4, 4., 4.25, 0, 0., -4, -4. -4.25):
+        for b in (2, 2., 2.5, -2, -2., -2.5):
+            assert divmod(S(a), S(b)) == dmod(a, b)
+    # done with Float/Integer now Rational
+    for a in range(-20, 21, 10):
+           a = a/S(5)
+           for b in (-10, -5, 5, 10):
+                  b = b/S(5)
+                  assert __divmod__(a, b) == dmod(a, b)
 
```
```suggestion

```

smichr committed:
remove redundant test

smichr commented:
> t would be a lot of work, so I don't blame anyone if no one wants to actually do it, but it would be interesting to compare this side-by-side with second pull request implementing the alternative idea of making floats and rationals always compare equal.

Your comment is old, now, @asmeurer, and I wonder if this is really worth pursuing. Here is my current sentiment:
1. we should not allow Float(1) == Integer(1) since they represent two different things; we are already operating in an environment where `x**.5 != sqrt(x)` so this change is not going to be that surprising.
2. we cannot assess precision when a user enters `x**.1` and our python dictate is to not guess when faced with ambiguity so we already leave it as `Float(.1)`
3. the user already has the option to sympify with the rational argument so `S(x**.1, rational=True) -> root(x, 10)`
4. an expression with Float can already be converted to Rational with *approximation* by using `nsimplify` though it is safer to use `replace` to target only Float: `expr.replace(lambda x:x.is_Float, lambda x: nsimplify(x))`
5. with this PR, an expression with Float can be converted *without approximation* to Rational using `expr.replace(lambda x:x.is_Float, lambda x: Rational(x))`

We don't have a special `is_odd` or `is_even` function because we just use `n % 2` to discriminate between the two; as I have shown in this PR's additional commits, we don't need an `is_int` to know if a number (Float or Rational) is free of a non-zero decimal because we can do `n % 1` to know what the decimal portion is (and the result will be integer 0 if there is no decimal portion).

@oscarbenjamin , you mentioned the need to detect 0.5 or 1/2. I didn't see that. What I saw was the need to discriminate against an expression that might have a 0.5 decimal portion in the wigner code. But we don't need to know if it's 0.5 we just need to know if it is not 0. If one *did* want to test for 0.5 or 1/2 then `(r:=n%1) and not (2*r)%1` would work.

Are there any final objections to this being committed without any additional helpers?

smichr commented:
What do you think about modifying `_aresame` to allow for `approx` testing of numbers. Perhaps this could be made into a Basic method to be used like `a.is_same(b)`, `a.is_same(b, approx=equal_valued)`, etc...
```python
def _aresame(a, b, approx=None):
    """Return True if a and b are structurally the same, else False.
    If `approx` is supplied, it will be used to test whether two
    numbers are the same or not. By default, only numbers of the
    same type will compare equal, so S.Half != Float(0.5).

    Examples
    ========

    In SymPy (unlike Python) two numbers do not compare the same if they are
    not of the same type:

    >>> from sympy import S
    >>> 2.0 == S(2)
    False
    >>> 0.5 == S.Half
    False

    By supplying a function with which to compare two numbers, such
    differences can be ignored. e.g. `equal_valued` will return True
    for decimal numbers having a denominator that is a power of 2,
    regardless of precision.

    >>> from sympy import Float
    >>> from sympy.core.basic import _aresame
    >>> from sympy.core.numbers import equal_valued
    >>> _aresame(S.Half/4, Float(0.125, 1), equal_valued)
    True
    >>> _aresame(Float(1, 2), Float(1, 10), equal_valued)
    True

    But decimals without a power of 2 denominator will compare
    as not being the same.

    >>> _aresame(Float(0.1, 9), Float(0.1, 10), equal_valued)
    False
    >>> import math
    >>> _aresame(Float(0.1, 9), Float(0.1, 10), math.isclose)
    True

    Other objects might compare the same even though types are not the
    same. This routine will only return True if two expressions are
    identical in terms of class types.

    >>> from sympy import eye
    >>> eye(1) == S(eye(1))  # mutable vs immutable
    True
    >>> _aresame(eye(1), S(eye(1)))
    False

    """
    from .numbers import Number
    from .function import AppliedUndef, UndefinedFunction as UndefFunc
    if isinstance(a, Number) and isinstance(b, Number):
        if approx:
            return approx(a, b)
        return a == b and a.__class__ == b.__class__
    for i, j in zip_longest(_preorder_traversal(a), _preorder_traversal(b)):
        if i != j or type(i) != type(j):
            if approx and isinstance(a, Number) and isinstance(b, Number):
                return approx(i, j)
            if ((isinstance(i, UndefFunc) and isinstance(j, UndefFunc)) or
                (isinstance(i, AppliedUndef) and isinstance(j, AppliedUndef))):
                if i.class_key() != j.class_key():
                    return False
            else:
                return False
    return True
```

smichr committed:
aresame -> Basic.is_same

smichr committed:
remove online edit indentation spaces

smichr commented on sympy/core/basic.py:
```diff
@@ -2052,6 +2052,79 @@ def _sage_(self):
+        same. This routine will only return True if two expressions are
+        identical in terms of class types.
+
+        >>> from sympy import eye
+        >>> eye(1) == S(eye(1))  # mutable vs immutable
+        True
+        >>> eye(1).is_same(S(eye(1)))
```
```suggestion
        >>> from sympy import eye, Basic
        >>> eye(1) == S(eye(1))  # mutable vs immutable
        True
        >>> Basic.is_same(eye(1), S(eye(1)))
```

smichr committed:
Update sympy/core/basic.py

smichr commented:
I do not see why this substitution is failing:
```python
>>> ExprCond(1,x<0).subs(x<0,x<0.)
(1, x < 0.0)
>>> Piecewise((1,x<0)).subs(x<0,x<0.)
Piecewise((1, x < 0))
```

oscarbenjamin commented:
> If 2.0 represents `2 +/- eps` then how can we ever evaluate an inequality of which it is part if we don't know the size of `eps`?

I think that this misses the point of this PR which is specifically about `==`. The other cases like `Eq(x, y)` or `x < y` etc are *semantic* statements. However in SymPy `==` is for *structural* equality. Comparing Float and Rational semantically is an entirely different (much larger) can of worms compared to the simple case of comparing Float and Rational structurally: a Float is always *structurally* distinct from a Rational.

oscarbenjamin commented:
> I do not see why this substitution is failing:
> 
> ```python
> >>> ExprCond(1,x<0).subs(x<0,x<0.)
> (1, x < 0.0)
> >>> Piecewise((1,x<0)).subs(x<0,x<0.)
> Piecewise((1, x < 0))
> ```

Presumably something in `Piecewise._eval_subs` does something that prevents this from working. It works in general for expression heads that don't define `_eval_subs`:
```python
In [4]: Basic(x < 0).subs(x < 0, x < 0.0)
Out[4]: Basic(x < 0.0)

In [5]: Basic(x < 0.0).subs(x < 0.0, x < 0)
Out[5]: Basic(x < 0)
```

smichr commented:
> Comparing Float and Rational semantically is an entirely different (much larger) can of worms

The problem is that this PR is opening that can because now we must answer semantical questions like whether `And(x<2, x<2.0)` should simplify to `x < 2` or `x < 2.0` (which is decided by comparing the numbers).

oscarbenjamin commented:
> we must answer semantical questions like whether `And(x<2, x<2.0)` should simplify to `x < 2` or `x < 2.0`

With master I get:
```python
In [1]: And(x < 2, x < 2.0)
Out[1]: x < 2.0 ∧ x < 2

In [2]: And(x < 2.0, x < 2)
Out[2]: x < 2.0 ∧ x < 2
```
With the PR I get the same. I don't see how anything is changed...

asmeurer commented:
> we must answer semantical questions like whether And(x<2, x<2.0) should simplify to x < 2 or x < 2.0

The spirit of this PR is that it shouldn't simplify. And anyway there's no issue mathematically with having redundant conditions in an And.

smichr commented:
> I don't see how anything is changed...

See the failing Piecewise test which uses an `And` like this; it was a recently added test which causes this PR to fail when merged.

smichr commented:
> The spirit of this PR is that it shouldn't simplify

This is only the tip of the "ice block of unknown dimensions". The Piecewise test that broke when merged with this PR shows the problems that will now be faced when trying to simplify Piecewise expressions. @oscargus has done a lot of work on simplification of inequalities. He might be able to comment on the new challenges that simplification will face. Maybe I am too paranoid. Here is the failing test:
```python
>>> x_d = Symbol('x_d')
>>> fun_sum = lambda x, a1, a2: Piecewise((0, x<a1),(1, x>=a1)) + Piecewise((0, x<a2),(1, x>=a2))
>>> eq = fun_sum((x_d), 1.0, 1); eq
Piecewise((0, x_d < 1), (1, True)) + Piecewise((0, x_d < 1.0), (1, True))
>>> integrate(eq, (x_d, -float('Inf'), x)) == 2*x - 2*Min(1, x)
False
>>> piecewise_fold(eq)
Piecewise((0, (x_d < 1.0) & (x_d < 1)), (1, (x_d < 1.0) | (x_d < 1)), (2, True))

>>> simplify(_)  # in master but not in this PR
Piecewise((0, x_d < 1), (2, True))
```

oscarbenjamin commented:
So on master we have:
```python
In [4]: p = Piecewise((0, (x < 1.0) & (x < 1)), (1, (x < 1.0) | (x < 1)), (2, True))

In [5]: p
Out[5]: 
⎧0  for x < 1.0 ∧ x < 1
⎪                      
⎨1  for x < 1.0 ∨ x < 1
⎪                      
⎩2       otherwise     

In [6]: simplify(p)
Out[6]: 
⎧0  for x < 1
⎨            
⎩2  otherwise
```
With the PR we have:
```python
In [3]: simplify(p)
Out[3]: 
⎧0  for x < 1.0
⎨              
⎩2   otherwise 
```
I'm not sure how exactly the simplification code depended on `==` for the PR to cause this change but I prefer the PR output for this particular case. If the original expression had floating point thresholds then I would prefer for the simplified expression to have floating point thresholds as well: inexact expressions should not be simplified into exact expressions.

smichr committed:
update is_same

smichr committed:
quick exit for _localzeros when no args

smichr committed:
add test for Min,Max

If 1.0 is imprecise then we can't say whether it is more or less than 1

smichr committed:
Update test_integrals.py 20781

Not sure that Min(0, 0.0) can be fixed to not evaluate since frozenset(Float(0), S.Zero) is {S.Zero}

smichr commented on sympy/integrals/tests/test_integrals.py:
```diff
@@ -2089,14 +2090,23 @@ def test_issue_20782():
-    assert integrate(-f, (x, -oo, 1)) == -1
+    assert integrate(f, (x, -oo, 1)) == 1.
+    assert integrate(-f, (x, -oo, 1)) == -1.
+
+
+def test_issue_20781():
+    x_d = Symbol('x_d')
+    fun_sum = lambda x, a1, a2: Piecewise((0, x<a1),(1, x>=a1)) + Piecewise((0, x<a2),(1, x>=a2))
+
+    assert integrate(fun_sum((x_d), 0, 0.0), (x_d, -float('Inf'), x)) == 2*x - 2*Min(0, x)
+    assert integrate(fun_sum((x_d), 0.0, 0), (x_d, -float('Inf'), x)) == 2*x - 2*Min(0, x)
+    assert integrate(fun_sum((x_d), 1.0, 1), (x_d, -float('Inf'), x)) == 2*x - Min(1.0, x) - Min(x, Max(1.0, 1))
```
```suggestion
```

smichr committed:
Update sympy/integrals/tests/test_integrals.py

smichr committed:
Update basic.py

smichr commented on sympy/functions/elementary/miscellaneous.py:
```diff
@@ -599,6 +599,8 @@ def _is_connected(cls, x, y):
         """
         Check if x and y are connected somehow.
         """
+        if not x.arg:
```
```suggestion
        if not x.args:
```

smichr committed:
Update sympy/functions/elementary/miscellaneous.py

asmeurer commented:
> Comparing Float and Rational semantically is an entirely different (much larger) can of worms compared to the simple case of comparing Float and Rational structurally: a Float is always structurally distinct from a Rational.

Maybe I'm missing something here, but I think semantic comparisons (inequalities) with Floats should be very straightforward. A Float should compare the same as its rational equivalent. As far as I can tell (from `Float._Frel`) this how it already works in master.

oscarbenjamin commented:
> Maybe I'm missing something here, but I think semantic comparisons (inequalities) with Floats should be very straightforward.

The questions would be more like how to handle things like `Eq(1, 1.0)` or `(x <= 1) & (x <= 1.0)` etc. Even with strict inequalities it gets thorny once rounding errors are involved like `(x < 1) & (x < 1 +- 1e-20)` and so on. In the structural equality case there is no ambiguity.

smichr commented:
>  A Float should compare the same as its rational equivalent

It's easy to make `==` be strict (that's what this PR does). But then what constitutes a "comparison"?

Do you mean you want the following?
* `Float(1) == S(1)` to be False  
* `Eq(Float(1), S(1))` to be True
* `Eq(.5, S.Half)` to be True
* `Eq(.1, S(1)/10)` to be False because all Floats have power-of-2 denominators?

Should comparisons (Eq-like) or equalities (`==`) be used to make decisions like whether to collapse `And(x < 1, x < 1.0)` to a single relationship? or `Min(1, 1.0)` to a single arg?

Does `Float(0)*1` evaluate to `S(0)` or `Float(0)`? If the former, then `Float(0) != S(0) but `Float(0)*1 == S(0)*1` and that seems odd.

smichr commented:
continued in #25614 

OK to close this, @oscarbenjamin ?

oscarbenjamin commented:
Continued in https://github.com/sympy/sympy/pull/25614
